### ISSUE TITLE
Need to upgrade DukTape to 1.3.0 for binary array object support.

### ISSUE THREAD

**Issue by Type1J** [AtomicGameEngine/AtomicGameEngine#371]
I might be wrong, and please correct me if I am.

Say that I want to pass binary data to JavaScript from C++ using SendEvent:
    VariantMap eventData;
    eventData.Insert(MakePair(StringHash("data"), Variant(PODVector<unsigned char>((const unsigned char *)ptr, length))));
    SendEvent("binary_data", eventData);

I would expect that there would be some type that I could get a JavaScript typed array or ArrayBuffer to be available from the eventData object, but this causes the "data" field to be undefined when the event is handled from JavaScript. Changing the PODVector to a String works as expected, but I'm assuming that strings are some form of UTF (might misrepresent binary data), and I want a byte array (no chance of misrepresentation).

**Comment by JoshEngebretson** [comment]
Here's where the supported variant types are pushed onto the stack: https://github.com/AtomicGameEngine/AtomicGameEngine/blob/master/Source/AtomicJS/Javascript/JSAPI.cpp#L406

What could happen here is to catch a vector and copy it to a buffer: http://duktape.org/guide.html#bufferobjects

**Comment by Type1J** [comment]
Yes, I believe a copy should happen here unless you know something that I don't, and you can somehow pull off a DataView instead of an ArrayBuffer to avoid the copy.

I have no preference on what type you'd want to catch. It seems that the most appropriate types are templated in C++. I **don't** want to get carried away by saying that each of the typed arrays in JavaScript should have a PODVector template type in C++. I think that a byte array type alone should be fine for now.

**Comment by Type1J** [comment]
Maybe PODVector templates *are* the best way to represent each of the typed arrays, but I'm not sure what would differentiate an ArrayBuffer from a Typed Uint8Array, though. I'm also not sure if I want this problem solved by a copy operation.

I know that this problem as it's worded in the title might still need to be solved, but maybe it could be sidestepped, and me made more efficient at the same time: We could make a new object that could be passed around by pointer in JavaScript (inherits from Atomic::Object), but stores some sort of heap allocated byte array that it can convert (by copying) to a typed array (with either endian) or a string (with a UTF encoding), appended, or passed back to C++ through a variant that doesn't copy its data, but only holds a reference to the Atomic::Object.

I think that we could treat it more like it's a file handle in JavaScript, not really an array, although in C++ (to make it efficient) it could expose a way t get a pointer to the data. I believe that Atomic/IO/MemoryBuffer already satisfies most of these requirements. What's needed is a way to pass it around in JavaScript (since it's not in the bindings due to it not being an Atomic::Object), and functions on the JavaScript side for getting data out of it in the various typed array and string forms.

Should this be a new issue (to allow this issue to track Variant to typed array conversions), or should this one be changed to something such as "Enable MemoryBuffer in JavaScript"?

To summarize: We need a new object that inherits from Atomic::Object and Atomic::IO::MemoryBuffer with some extra bits on the JavaScript side to (tediously by way of copy) convert the contents to and from typed arrays and string types.

**Comment by Type1J** [comment]
It doesn't have to be MemoryBuffer. It could be SharedArrayPtr<unsigned char> maybe?

**Comment by Type1J** [comment]
I just found Atomic::VectorBuffer, which seems to be a really good fit. I'll try to isolate the changes into a pull request as soon as I can.

**Comment by JoshEngebretson** [comment]
Sorry for going MIA on this... I am a bit snowed under working on cleaning up the build, CI backend, and getting the Editor binaries code signed so Windows 10 doesn't refuse to run it :)

If you can spare a little JS usage snippet that would be great to see.

**Comment by Type1J** [comment]
I understand :-). I'm just trying to make sure that binary messages sent by websockets actually show up in the JavaScript version of the event data. Also, in http requests, I'm trying to make sure that you could get binary data that the string implementation would reinterpret as UTF8 if you want it. I'll show an example usage when I'm done.

**Comment by JoshEngebretson** [comment]
A slug of build updates are now on master, let me know if you run into problems updating :)

http://atomicgameengine.com/forum/#/discussion/148/recent-updates-on-master-branch-nov-2nd-2015

- Josh

**Comment by Type1J** [comment]
I changed the title because I need DukTape 1.3.0 at least to get support for things like duk_push_buffer_object(), and to get support for the ArrayBuffer APIs.

@JoshEngebretson, can these libraries be upgraded to at least 1.3.0 in the AGE repo? (In other words: is there a reason that they haven't been upgraded?) It looks like the last update was just before 1.3.0, but we don't seem to have duk_push_buffer_object(), which DukTape docs says was added in 1.3.0.

**Comment by JoshEngebretson** [comment]
@Type1J I updated to Duktape master this morning:

#416 

Enjoy!  Looking forward to seeing what you are cooking up :)

**Comment by Type1J** [comment]
Thanks! I hope it meets or exceeds your expectations. :-)

**Comment by Type1J** [comment]
Between #450 and #453, we're good here. :-)
